File: docker/worker-customcode-python/harness/honua_customcode_harness/sourcefetch.py
"""Clone + pin user code at an exact git SHA, then verify the checkout.

The harness only ever runs code at a fully-pinned 40-hex SHA. After fetching we
``git rev-parse HEAD`` and assert it equals the requested SHA — so a malicious
or misconfigured remote that resolves a ref to a different commit fails hard
instead of silently running unexpected code.
"""

from __future__ import annotations

import subprocess
from collections.abc import Sequence
from pathlib import Path

from .jobspec import is_valid_git_sha


class SourceFetchError(RuntimeError):
    """Raised when cloning/checkout/verification of user code fails."""


def _run_git(args: Sequence[str], *, cwd: Path | None = None, runner=subprocess.run) -> str:
    """Run a git command, returning stdout. Raises SourceFetchError on failure.

    ``runner`` is injectable so tests can stub git without a real binary.
    """
    cmd = ["git", *args]
    proc = runner(
        cmd,
        cwd=str(cwd) if cwd else None,
        capture_output=True,
        text=True,
        check=False,
    )
    if proc.returncode != 0:
        raise SourceFetchError(
            f"git {' '.join(args)} failed ({proc.returncode}): {proc.stderr.strip()}"
        )
    return (proc.stdout or "").strip()


def clone_pinned(
    repo_url: str,
    git_ref: str,
    dest: Path,
    *,
    runner=subprocess.run,
) -> Path:
    """Clone ``repo_url`` and check out exactly ``git_ref`` (a 40-hex SHA).

    Strategy: shallow ``init`` + ``fetch --depth 1 <sha>`` + ``checkout <sha>``.
    A direct fetch of the SHA keeps the download minimal and works whether or not
    the host allows fetch-by-sha (we fall back to a shallow clone if it does
    not). Finally assert ``rev-parse HEAD == sha``.
    """
    if not is_valid_git_sha(git_ref):
        # Defense-in-depth: never pass an unvalidated ref to git.
        raise SourceFetchError(
            f"refusing to fetch non-SHA git_ref {git_ref!r} (must be 40-hex)."
        )

    dest.mkdir(parents=True, exist_ok=True)
    _run_git(["init", "--quiet"], cwd=dest, runner=runner)
    _run_git(["remote", "add", "origin", repo_url], cwd=dest, runner=runner)

    try:
        _run_git(
            ["fetch", "--depth", "1", "origin", git_ref],
            cwd=dest,
            runner=runner,
        )
    except SourceFetchError:
        # Some hosts disable uploadpack.allowReachableSHA1InWant; fall back to a
        # shallow clone of the default branch + an unshallow-by-SHA fetch.
        _run_git(["fetch", "--depth", "1", "origin"], cwd=dest, runner=runner)
        _run_git(["fetch", "--depth", "50", "origin"], cwd=dest, runner=runner)

    _run_git(["checkout", "--quiet", "--detach", git_ref], cwd=dest, runner=runner)

    head = _run_git(["rev-parse", "HEAD"], cwd=dest, runner=runner)
    if head != git_ref:
        raise SourceFetchError(
            f"checkout verification failed: HEAD is {head!r} but expected {git_ref!r}."
        )
    return dest

File: docker/worker-customcode-dotnet/harness/src/Honua.CustomCode.Harness/SourceFetch.cs
// Copyright (c) Honua. All rights reserved.
// Licensed under the Elastic License 2.0. See LICENSE in the project root.

using System.Diagnostics;

namespace Honua.CustomCode.Harness;

/// <summary>Raised when cloning/checkout/verification of user code fails.</summary>
public sealed class SourceFetchException(string message) : Exception(message);

/// <summary>The outcome of running a process: exit code + captured stdout/stderr.</summary>
/// <param name="ExitCode">The process exit code.</param>
/// <param name="StdOut">Captured standard output.</param>
/// <param name="StdErr">Captured standard error.</param>
public sealed record ProcessResult(int ExitCode, string StdOut, string StdErr);

/// <summary>
/// Clone + pin user code at an exact git SHA, then verify the checkout. The harness
/// only ever runs code at a fully-pinned 40-hex SHA; after fetching we
/// <c>git rev-parse HEAD</c> and assert it equals the requested SHA. Mirrors the
/// Python harness's <c>sourcefetch.py</c>.
/// </summary>
public sealed class SourceFetch
{
    private readonly Func<IReadOnlyList<string>, string?, ProcessResult> _runner;

    /// <summary>Creates a source fetcher.</summary>
    /// <param name="gitRunner">An injectable git runner (for tests); defaults to invoking the real git binary.</param>
    public SourceFetch(Func<IReadOnlyList<string>, string?, ProcessResult>? gitRunner = null)
    {
        _runner = gitRunner ?? RunGit;
    }

    /// <summary>
    /// Clone <paramref name="repoUrl"/> and check out exactly <paramref name="gitRef"/>
    /// (a 40-hex SHA), then assert <c>rev-parse HEAD == sha</c>.
    /// </summary>
    /// <param name="repoUrl">The git repository URL.</param>
    /// <param name="gitRef">The pinned 40-hex commit SHA.</param>
    /// <param name="destination">The destination directory.</param>
    /// <returns>The destination directory.</returns>
    /// <exception cref="SourceFetchException">On any clone/checkout/verify failure.</exception>
    public string ClonePinned(string repoUrl, string gitRef, string destination)
    {
        if (!JobSpec.IsValidGitSha(gitRef))
        {
            // Defense-in-depth: never pass an unvalidated ref to git.
            throw new SourceFetchException($"refusing to fetch non-SHA git_ref '{gitRef}' (must be 40-hex).");
        }

        Directory.CreateDirectory(destination);
        Git(destination, "init", "--quiet");
        Git(destination, "remote", "add", "origin", repoUrl);

        try
        {
            // A direct fetch of the SHA keeps the download minimal.
            Git(destination, "fetch", "--depth", "1", "origin", gitRef);
        }
        catch (SourceFetchException)
        {
            // Some hosts disable uploadpack.allowReachableSHA1InWant; fall back to a
            // shallow clone of the default branch + a deeper fetch by SHA.
            Git(destination, "fetch", "--depth", "1", "origin");
            Git(destination, "fetch", "--depth", "50", "origin");
        }

        Git(destination, "checkout", "--quiet", "--detach", gitRef);

        var head = Git(destination, "rev-parse", "HEAD").Trim();
        if (!string.Equals(head, gitRef, StringComparison.Ordinal))
        {
            throw new SourceFetchException(
                $"checkout verification failed: HEAD is '{head}' but expected '{gitRef}'.");
        }

        return destination;
    }

    private string Git(string cwd, params string[] args)
    {
        var result = _runner(args, cwd);
        if (result.ExitCode != 0)
        {
            throw new SourceFetchException(
                $"git {string.Join(' ', args)} failed ({result.ExitCode}): {result.StdErr.Trim()}");
        }

        return result.StdOut ?? string.Empty;
    }

    private static ProcessResult RunGit(IReadOnlyList<string> args, string? cwd)
    {
        var psi = new ProcessStartInfo("git")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            WorkingDirectory = cwd ?? string.Empty,
        };
        foreach (var arg in args)
        {
            psi.ArgumentList.Add(arg);
        }

        using var process = Process.Start(psi)
            ?? throw new SourceFetchException("failed to start git.");
        var stdout = process.StandardOutput.ReadToEnd();
        var stderr = process.StandardError.ReadToEnd();
        process.WaitForExit();
        return new ProcessResult(process.ExitCode, stdout, stderr);
    }
}
